### PR TITLE
Add PT_REGS_PARMx_SYSCALL helpers for all archs

### PR DESCRIPTION
**Pull request by craig65535**

### CODE REVIEW

**Comment by craig65535** [comment]
@yonghong-song with this change, my understanding is that a BPF program would use the PT_REGS_PARM1 macro as follows, on x86:

```C
struct pt_regs *ctx2 = (struct pt_regs *)PT_REGS_PARM1(ctx);
bpf_probe_read(&parm1, sizeof(parm1), &PT_REGS_PARM1_SYSCALL(ctx2));
```

But, I'm wondering how I could make this work more portably on other architectures. Maybe it's better to define the macros as follows:
```C
#define PT_REGS_PARM4_SYSCALL(ctx) PT_REGS_PARM4((struct pt_regs *)PT_REGS_PARM1(ctx))
```

Other architectures with `!defined(CONFIG_ARCH_HAS_SYSCALL_WRAPPER)`, or `defined(__s390x__)` could use 
```C
#define PT_REGS_PARM4_SYSCALL(ctx) PT_REGS_PARM4(ctx)
```

And then a BPF program would only need to contain
```C
bpf_probe_read(&parm4, sizeof(parm4), &PT_REGS_PARM4_SYSCALL(ctx));
```

What do you think?

File: src/cc/export/helpers.h [hunk 1]
```diff
@@ -1075,45 +1075,67 @@ int bpf_usdt_readarg_p(int argc, struct pt_regs *ctx, void *buf, u64 len) asm("l
...
 #define PT_REGS_PARM2(x)	((x)->regs[1])
+#define PT_REGS_PARM2_SYSCALL(ctx)	PT_REGS_PARM2(ctx)
 #define PT_REGS_PARM3(x)	((x)->regs[2])
+#define PT_REGS_PARM3_SYSCALL(ctx)	PT_REGS_PARM3(ctx)
 #define PT_REGS_PARM4(x)	((x)->regs[3])
+#define PT_REGS_PARM4_SYSCALL(ctx)	PT_REGS_PARM4(ctx)
 #define PT_REGS_PARM5(x)	((x)->regs[4])
+#define PT_REGS_PARM5_SYSCALL(ctx)	PT_REGS_PARM5(ctx)
 #define PT_REGS_PARM6(x)	((x)->regs[5])
+#define PT_REGS_PARM6_SYSCALL(ctx)	PT_REGS_PARM6(ctx)
```
Comment (yonghong-song) [inline src/cc/export/helpers.h:1138]: A lot of duplication here, can we do something at the end of all `PT_REGS_PARM*`? Something like
```
#define PT_REGS_PARM1_SYSCALL(ctx)	PT_REGS_PARM1(ctx)
#define PT_REGS_PARM2_SYSCALL(ctx)	PT_REGS_PARM2(ctx)
......
#ifdef bpf_target_x86
#define PT_REGS_PARM4_SYSCALL(ctx)	((ctx)->r10)
#else 
#define PT_REGS_PARM4_SYSCALL(ctx)	PT_REGS_PARM4(ctx)
#endif 
...
#ifdef PT_REGS_PARM6
#define PT_REGS_PARM6_SYSCALL(ctx)	PT_REGS_PARM6(ctx)
#endif 
```

**Review by yonghong-song — COMMENTED** [review]
@craig65535 I am thinking the following. For example

```
#if defined(CONFIG_ARCH_HAS_SYSCALL_WRAPPER) && !defined(__s390x__)
#define PT_REGS_PARM2_SYSCALL(ctx) \
  ({ struct pt_regs *ctx2 = (struct pt_regs *)PT_REGS_PARM1(ctx); PT_REGS_PARM1(ctx2); })
...
#else 
...
```
might work. The issue is it loads ctx2 every time for a parameter.

In my opinion, users should access syscall parameters through syscall__ prefix and the bcc will automatically take care of it. For advanced users, doing
```
struct pt_regs *ctx2 = (struct pt_regs *)PT_REGS_PARM1(ctx);
bpf_probe_read(&parm1, sizeof(parm1), &PT_REGS_PARM1_SYSCALL(ctx2));
```
maybe fine?

**Comment by craig65535** [comment]
> In my opinion, users should access syscall parameters through syscall__ prefix and the bcc will automatically take care of it. For advanced users, doing
> 
> ```
> struct pt_regs *ctx2 = (struct pt_regs *)PT_REGS_PARM1(ctx);
> bpf_probe_read(&parm1, sizeof(parm1), &PT_REGS_PARM1_SYSCALL(ctx2));
> ```
> 
> maybe fine?

This is more efficient, but again, is specific to architectures with syscall wrappers.

What do you think about introducing a new helper that returns the context required for syscall params?

```C
#if defined(CONFIG_ARCH_HAS_SYSCALL_WRAPPER) && !defined(__s390x__)
#  define PT_REGS_SYSCALL_CTX(ctx) ((struct pt_regs *)PT_REGS_PARM1(ctx))
#else
#  define PT_REGS_SYSCALL_CTX(ctx) (ctx)
#endif
```

And then BPF programs could write portable code:

```C
struct pt_regs *ctx2 = PT_REGS_SYSCALL_CTX(ctx);
bpf_probe_read(&parm1, sizeof(parm1), &PT_REGS_PARM1_SYSCALL(ctx2));
```

**Comment by yonghong-song** [comment]
I like the idea for PT_REGS_SYSCALL_CTX. Seems like good compromise, but still hide the complexity with CONFIG options.

**Comment by craig65535** [comment]
@yonghong-song PTAL

**Comment by yonghong-song** [comment]
[buildbot, test this please]

**Comment by craig65535** [comment]
@yonghong-song I'm not sure why the build failed. Looks like it failed in fc26 only.

**Comment by yonghong-song** [comment]
[buildbot, test this please]

**Comment by yonghong-song** [comment]
> @yonghong-song I'm not sure why the build failed. Looks like it failed in fc26 only.

a flaky test, just restarted the test.

**Pull request merged**